### PR TITLE
Added check for duplicate Registry key paths #41

### PR DESCRIPTION
**Pull request by joachimmetz**
[Review on Reviewable](https://reviewable.io/reviews/forensicartifacts/artifacts/145)

### CODE REVIEW

File: tools/validator.py [hunk 1]
```diff
@@ -44,10 +48,23 @@ def CheckFile(self, filename):
...
 
-        self.defined_artifact_names.add(artifact_definition.name)
+        self._defined_artifact_names.add(artifact_definition.name)
         for source in artifact_definition.sources:
-          if isinstance(source, source_type.ArtifactSourceType):
-            self.artifact_name_references.update(source.names)
+          # TODO: change to ARTIFACT_GROUP
+          if source.type_indicator == definitions.TYPE_INDICATOR_ARTIFACT:
+            self._artifact_name_references.update(source.names)
+
```
Comment (destijl) [inline tools/validator.py]: # Detect duplicate registry key paths

File: tools/validator.py [hunk 2]
```diff
@@ -44,10 +48,23 @@ def CheckFile(self, filename):
...
+          elif source.type_indicator in (
+              definitions.TYPE_INDICATOR_WINDOWS_REGISTRY_KEY):
+            intersection = self._artifact_registry_key_paths.intersection(
+                set(source.keys))
+            if intersection:
+              duplicate_key_paths = u'\n'.join(intersection)
+              logging.warning((
```
Comment (destijl) [inline tools/validator.py]: I think we should set result=False here. Is there a case where we want to have an exact duplicate key? Seems like we should restructure artifacts if that is the case. We should warn if we're doing fuzzy matching, but exact dups should cause an error.

File: tools/validator.py (hunk 1 above)
Comment (joachimmetz) [inline tools/validator.py]: I'll change this into a separate method so we don't need to add comments.

File: tools/validator.py (hunk 2 above)
Comment (joachimmetz) [inline tools/validator.py]: Ack

**Comment by joachimmetz** [comment]
Changed to _HasDuplicateRegistryKeyPaths

File: tools/validator.py [hunk 3]
```diff
@@ -21,15 +21,47 @@ class ArtifactDefinitionsValidator(object):
...
     self._artifact_registry = registry.ArtifactDefinitionsRegistry()
-    self.defined_artifact_names = set()
-    self.artifact_name_references = set()
+    self._artifact_registry_key_paths = set()
+    self._defined_artifact_names = set()
+
+  def _HasDuplicateRegistryKeyPaths(
+      self, filename, artifact_definition, source):
+    """Checks if Registry key paths are not already defined by other artifacts.
+
```
Comment (destijl) [inline tools/validator.py]: Add comment: `Note this will only catch exact path duplicates.`

**Comment by destijl** [comment]
LGTM, minor comment.

**Comment by joachimmetz** [comment]
I've added it to the docstring instead

**Comment by destijl** [comment]
Ah yep that's what I meant :)

**Pull request merged**